Feat/soft delete

lesagi opened the pull request:
Hi,

I added the option for soft for both InMemory and Relational, need to add Redis

I made it optional, and also I prevent getting the URL if the URL was soft deleted, please let me know what you think

I've merged the release branch to prevent conflicts in the future, that made the PR a bit hard to follow, my new Impl starts at:

"Merge remote-tracking branch 'origin/release/0.2.0' into release/0.2.0"

#79

lesagi committed:
Add the option to set a custom_id for users with admin right only

lesagi committed:
Merge branch 'release/0.2.0' of https://github.com/mini-services/miniurl into release/0.2.0

SnirShechter committed:
Merge branch 'release/0.2.0' of github.com:SnirShechter/miniurl into sagi/release/0.2.0

lesagi committed:
Update src/services/storage/drivers/relational/index.ts

destructuring function argument

Co-authored-by: Snir Shechter <me@snir.sh>

lesagi committed:
Update src/routes/url.ts

Co-authored-by: Snir Shechter <me@snir.sh>

lesagi committed:
Update src/services/storage/drivers/relational/index.ts

Co-authored-by: Snir Shechter <me@snir.sh>

lesagi committed:
Update src/services/storage/drivers/relational/index.ts

Co-authored-by: Snir Shechter <me@snir.sh>

lesagi committed:
Unauthorized Error on Custom ID
throwing an error when non-privileged user trying to create a custom id

lesagi committed:
Use postgres Error number to identify id duplicates

lesagi committed:
Check if custom_id already exists before saving

lesagi committed:
Merge remote-tracking branch 'origin/release/0.2.0' into release/0.2.0

lesagi committed:
Add 'deletedAt' property to StoredUrl type

lesagi committed:
Replace hard delete with soft delete for InMemory Driver

lesagi committed:
Store ISOString in the 'deletedAt' field instead of UTCString

lesagi committed:
Prevent unauthorized user to fetch soft deleted url

lesagi committed:
Make the hard delete optional along with soft delete

lesagi committed:
Add the option for soft delete for Relational DB

lesagi committed:
Relational: Prevent unauthorized user to fetch soft deleted url

SnirShechter commented on src/services/storage/drivers/relational/index.ts:
```diff
@@ -69,15 +70,15 @@ export class RelationalStorage implements StorageDriver {
+				.select('*')
+				.where('id', id)
+				.first()
+			if (storedUrl?.deletedAt && !options.isAuthorized) throw new UnauthorizedError();
```
If a user asks for a URL with id=4 (doesn't exist), it'll receive a 404. If they request a URL with id=3 (deleted), they'll receive an UnauthorizedError. This gives away the fact we use soft-delete instead of hard-delete (bad).

lesagi replied:
Changed both to 404

SnirShechter commented on src/services/storage/drivers/relational/index.ts:
```diff
@@ -93,9 +94,14 @@ export class RelationalStorage implements StorageDriver {
+			if (options.hardDelete) {
+				await this.storage.db.table<StoredUrl>('urls').where('id', id).delete()
+			} else {
+				await this.storage.db.table<StoredUrl>('urls').update({ deletedAt: new Date().toISOString() }).where('id', id)
```
I think it'll give the locale date (not GMT). Can you check?

SnirShechter commented on src/services/storage/drivers/relational/index.ts:
```diff
@@ -93,9 +94,14 @@ export class RelationalStorage implements StorageDriver {
+			}
 		}
 
 		public async deleteOverdue(timespanMs: number): Promise<number> {
```
What about deleteOverdue? We need it to soft-delete too.

lesagi replied:
added a deleteOverdue + migration to support the soft delete in postgres

Danielg212 reviewed (commented):
@lesagi please see Snir comments

lesagi committed:
Update Soft Delete

- Add a migration for the new 'deletedAt' Column

- Define a default SOFT delete method for deleteOverdue

- Change the 'options' parameter passed to the drivers' functions

lesagi commented on src/services/storage/drivers/relational/index.ts:
```diff
@@ -69,15 +70,15 @@ export class RelationalStorage implements StorageDriver {
 		constructor(public storage: RelationalStorage) { }
 
-		public async get(id: string, options = { withInfo: false }): Promise<StoredUrl | UrlWithInformation> {
+		public async get(id: string, options = { withInfo: false, isAuthorized: false }): Promise<StoredUrl | UrlWithInformation> {
```
Took the name you suggested

lesagi commented:
> @lesagi please see Snir comments

Hi, Snir was updated yesterday, it was WIP, not can be reviewed again, thanks for the heads up!

lesagi committed:
Remove unecessary import

lesagi committed:
Refactor softDelete function location in InMemoryStorage

lesagi committed:
Add a call to storage.softDelete in Overdue
